pkg/deployment/ipam.go

```diff
@@ -0,0 +1,242 @@
+					// Add domain once IPset status includes it
+					fqdnName := strings.Join([]string{hostName,
+						strings.ToLower(string(res.Network))}, ".")
+					fqdnNames = append(fqdnNames, fqdnName)
+					dnsRecord.Hostnames = fqdnNames
+					allDNSRecords = append(allDNSRecords, dnsRecord)
+				}
+			}
+		}
+	}
+	util.LogForObject(helper, "Reconciling DNSData", instance)
+	dnsData := &networkv1beta1.DNSData{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: instance.Namespace,
+			Name:      instance.Name,
+		},
+	}
+	_, err := controllerutil.CreateOrPatch(ctx, helper.GetClient(), dnsData, func() error {
+		dnsData.Spec.Hosts = allDNSRecords
+		// Set controller reference to the DataPlaneNode object
+		err := controllerutil.SetControllerReference(
+			helper.GetBeforeObject(), dnsData, helper.GetScheme())
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+// EnsureDNSData Ensures DNSData is created
+func EnsureDNSData(ctx context.Context, helper *helper.Helper,
+	instance *dataplanev1beta1.OpenStackDataPlaneRole,
+	nodes *dataplanev1beta1.OpenStackDataPlaneNodeList) (bool, error) {
+	// Verify dnsmasq CR exists
+	dnsmasq := &networkv1beta1.DNSMasq{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "dnsmasq",
+			Namespace: instance.Namespace,
+		},
+	}
+	key := client.ObjectKeyFromObject(dnsmasq)
+	if err := helper.GetClient().Get(ctx, key, dnsmasq); err != nil {
+		if k8s_errors.IsNotFound(err) {
+			util.LogForObject(helper, "No dnsmasq running, DNSData won't be created", instance)
+			return true, nil
+		}
+		return false, err
+	}
+	// TODO(rabi) Get dnsmasq LB IP and update BaremetealSpec
+	// and ansible inventory.
+
+	// Create or Patch DNSData
+	err := createOrPatchDNSData(ctx, helper, instance, nodes)
+	if err != nil {
+		instance.Status.Conditions.MarkFalse(
+			dataplanev1beta1.RoleDNSDataReadyCondition,
+			condition.ErrorReason, condition.SeverityError,
+			dataplanev1beta1.RoleDNSDataReadyErrorMessage)
+		return false, err
+	}
+
+	dnsData := &networkv1beta1.DNSData{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.Name,
+			Namespace: instance.Namespace,
+		},
+	}
+	key = client.ObjectKeyFromObject(dnsData)
+	err = helper.GetClient().Get(ctx, key, dnsData)
+	if err != nil {
+		return false, err
+	}
+
+	if !dnsData.IsReady() {
+		util.LogForObject(helper, "DNSData not ready yet waiting", instance)
+		return false, nil
+	}
+	instance.Status.Conditions.MarkTrue(
+		dataplanev1beta1.RoleDNSDataReadyCondition,
+		dataplanev1beta1.RoleDNSDataReadyMessage)
+	return true, nil
+}
+
+// reserveIPs Reserves IPs by creating IPSets
+func reserveIPs(ctx context.Context, helper *helper.Helper,
+	instance *dataplanev1beta1.OpenStackDataPlaneRole,
+	nodes *dataplanev1beta1.OpenStackDataPlaneNodeList, allIPSets []networkv1beta1.IPSet) error {
+
+	// Verify NetConfig CR exists
+	netConfig := &networkv1beta1.NetConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "netconfig",
```

Comment:
Yeah possible, In an earlier PS I had a check to list all ipsets and if the length is zero assumed that IPAM can be used. I can change to that. What other ways you can check? We have to make sure that this works both when netconfig (whatever the name is) there or not. I assumed that we would have a fixed name for netconfig like other services.

Reply:
In general I think you could just remove it, as we get the check with https://github.com/openstack-k8s-operators/infra-operator/pull/61 , otherwise we should do the List. The netconfig is something the user creates, so we can not expect a fixed name.